### DISCUSSION TITLE
Opening choice drop down menu blocks GUI from updating

### DISCUSSION THREAD

**Discussion by schmauder** [fltk-rs/fltk-rs#1619] [Q&A] [answered]
I use the channel method to handle events, but there is an issue. When pressing a choice widget the drop down menu opens and as long as the drop down menu is opened all widgets are not updated.

Attached is example code: There is a second thread that creates a "Tick" event every second and sends it through the channel. In the event loop the tick event is handled and a timer is incremented. When clicking the choice widget the timer visibly stops incrementing. But after selecting a choice element the timer moves on indicating that the event loop is not blocked (if leaving the choice menu open for e.g. 10s, the timer moves on 10s later). Can this blocking be prevented in an easy way?

    use fltk::prelude::*;

    #[derive(Clone)]
    enum Event {
        Tick,
        Choice,
    }

    fn main() {
        let app = fltk::app::App::default().with_scheme(fltk::app::Scheme::Gtk);

        let mut wind = fltk::window::Window::default()
            .with_size(320, 240)
            .center_screen();

        let mut timer = fltk::frame::Frame::default()
            .with_size(100, 40)
            .center_of(&wind);
        timer.set_label("0");

        let mut choice = fltk::menu::Choice::default()
            .with_size(100, 40)
            .below_of(&timer, 10);
        choice.add_choice("Choice 1");
        choice.add_choice("Choice 2");
        choice.set_value(0);

        wind.end();
        wind.show();

        let (sender, receiver) = fltk::app::channel();
        choice.emit(sender.clone(), Event::Choice);

        std::thread::spawn(move || loop {
            std::thread::sleep(std::time::Duration::from_secs(1));
            sender.send(Event::Tick);
        });

        let mut counter = 0;    

        while app.wait() {
            if let Some(event) = receiver.recv() {
                match event {
                    Event::Tick => {
                        counter += 1;
                        timer.set_label(&counter.to_string());
                    },
                    Event::Choice => (),
                }
            }
        }
    }

**Comment by MoAlyousef** [comment, marked as answer]
Hi
Thanks for the report. `while app.wait()` is not really suitable since it's FLTK event driven. I came to find this out also rather late.
If you have messages from another thread for example, or that are not initiated by actual FLTK events, it's more appropriate to install an idle handler:
```rust
use fltk::prelude::*;

#[derive(Clone)]
enum Event {
    Tick,
    Choice,
    Stop,
}

fn main() {
    let app = fltk::app::App::default().with_scheme(fltk::app::Scheme::Gtk);

    let mut wind = fltk::window::Window::default()
        .with_size(320, 240)
        .center_screen();

    let mut timer = fltk::frame::Frame::default()
        .with_size(100, 40)
        .center_of(&wind);
    timer.set_label("0");

    let mut choice = fltk::menu::Choice::default()
        .with_size(100, 40)
        .below_of(&timer, 10);
    choice.add_choice("Choice 1");
    choice.add_choice("Choice 2");
    choice.set_value(0);

    let mut stop = fltk::button::Button::default()
        .with_label("Stop")
        .with_size(100, 40)
        .below_of(&choice, 10);

    wind.end();
    wind.show();

    let (sender, receiver) = fltk::app::channel();
    choice.emit(sender.clone(), Event::Choice);
    stop.emit(sender.clone(), Event::Stop);

    std::thread::spawn(move || loop {
        std::thread::sleep(std::time::Duration::from_secs(1));
        sender.send(Event::Tick);
    });

    let mut counter = 0;

    fltk::app::add_idle3(move |h| {
        if let Some(event) = receiver.recv() {
            match event {
                Event::Tick => {
                    counter += 1;
                    timer.set_label(&counter.to_string());
                }
                Event::Stop => fltk::app::remove_idle3(h),
                Event::Choice => (),
            }
        }
    });

    app.run().unwrap();
}
```

The remove_idle3 that's used here is just for demonstration, it doesn't stop the thread loop for example. Also this uses a lot of cpu, so you would need a sleep call inside the idle handler.
Another option is to use a timeout or just set the timer's (Frame) label inside the thread loop.

**Reply by schmauder** [reply]
Hi MoAlyousef,
thanks a lot for the solution and the explanations. Replacing `app::wait()` with `app::add_idle3` works and i will go on using that. As you described without using `thread::sleep` one CPU will be saturated at 100%.

While not having tried a solution with timeout, i tried to update the frame in the tick loop directly and leaving the `app::wait()` function in (see code below). This doesn't do the thing i want but it makes more clear what `app::wait()` does. As long as no event occurs (like moving the mouse) the GUI will not be updated. Seems like i need a `app::run()` and just using `app::wait()` comes with some restrictions.

```
        let (sender, receiver) = fltk::app::channel();
        choice.emit(sender.clone(), Event::Choice);

        let mut timer2 = timer.clone();
        let mut counter = 0;    
        std::thread::spawn(move || loop {
            timer2.set_label(&counter.to_string());
            counter += 1;
            std::thread::sleep(std::time::Duration::from_secs(1));
        });

        while app.wait() {
            if let Some(event) = receiver.recv() {
                match event {
                    Event::Choice => timer.set_label("Selected"),
                }
            }
        }
    }
```

**Reply by MoAlyousef** [reply]
You can modify the thread loop like this:
```rust
    std::thread::spawn(move || loop {
        timer2.set_label(&counter.to_string());
        counter += 1;
        std::thread::sleep(std::time::Duration::from_secs(1));
        fltk::app::awake();
    });
```

The `awake` call informs FLTK that another thread of execution has modified the gui.
